nemo/collections/nlp/models/language_modeling/megatron/bert/bert_model.py

```diff
@@ -154,6 +174,260 @@ def post_language_model_processing(
+
+        # Output
+        if self.post_process:
+            # TODO: Make sure you are passing in the mpu_vocab_size properly
+            self.lm_head = MCoreBertLMHead(
+                self.config.hidden_size,
+                self.config,
+                self.parallel_output,
+                self.vocab_size,
+                self.pre_process,
+                self.share_embeddings_and_output_weights,
+            )
+
+            self.output_layer = self.lm_head.output_layer
+
+            self.binary_head = None
+            if self.add_binary_head:
+                # TODO: Shoudl switch this to TE ?
+                self.binary_head = mcore_get_linear_layer(
+                    self.config.hidden_size, 2, self.config.init_method, self.config.perform_initialization
+                )
+
+    def forward(
+        self,
+        input_ids: Tensor,
+        attention_mask: Tensor,
+        tokentype_ids: Tensor = None,
+        lm_labels: Tensor = None,
+        inference_params=None,
+    ):
+        """Forward function of BERT model
+
+        Forward function of the BERT Model This function passes the input tensors
+        through the embedding layer, and then the encoder and finally into the post
+        processing layer (optional).
+
+        It either returns the Loss values if labels are given  or the final hidden units
+        """
+        hidden_states = super(MCoreBertModelWrapper, self).forward(
+            input_ids, attention_mask, tokentype_ids, lm_labels, inference_params
+        )
+        if not self.post_process:
+            return hidden_states
+
+        if self.add_pooler:
+            pooled_output = self.pooler(hidden_states, 0)
+
+        if self.return_embeddings:
+            embeddings = torch.transpose(hidden_states, 0, 1)
+            masks = torch.sum(attention_mask, dim=1)
+            # Collect masked embeddings.
+            output = torch.zeros(
+                size=(embeddings.shape[0], embeddings.shape[2]),
+                dtype=torch.float32,
+                device=torch.cuda.current_device(),
+            )
+            for i, (embedding, mask) in enumerate(zip(embeddings, masks)):
+                output[i, :] = torch.mean(embedding[1 : mask - 1], dim=0)
+            return output
+
+        # logits and loss
+        output_weight = None
+        if self.share_embeddings_and_output_weights:
+            output_weight = self.shared_embedding_or_output_weight()
+
+        logits = self.lm_head(hidden_states=hidden_states, word_embeddings_weight=output_weight)
+
+        binary_logits = None
+        if self.binary_head is not None and self.add_pooler:
+            binary_logits = self.binary_head(pooled_output)
+
+        if lm_labels is None:
+            # [s b h] => [b s h]
+            return logits.transpose(0, 1).contiguous(), binary_logits
+
+        loss = self.compute_language_model_loss(lm_labels, logits)
+
+        return loss, binary_logits
+
+
 class BertModel(MegatronModule):
```

Comment:
Can you add a log statement saying it will be deprecated.

Reply:
added: 

warnings.warn("NeMoBertModel will be deprecated mid 2024. Use MCoreBertModelWrapperWithPostLNSupport instead.", DeprecationWarning)